sklearn/neighbors/nca.py

```diff
@@ -0,0 +1,545 @@
+            elif init == 'pca':
+                pca = PCA(n_components=n_features_out,
+                          random_state=self.random_state_)
+                t_pca = time.time()
+                if self.verbose:
+                    print('Finding principal components... ', end='')
+                    sys.stdout.flush()
+
+                pca.fit(X)
+                if self.verbose:
+                    print('done in {:5.2f}s'.format(time.time() - t_pca))
+
+                transformation = pca.components_
+        return transformation
+
+    def _callback(self, transformation):
+        """Called after each iteration of the optimizer.
+
+        Parameters
+        ----------
+        transformation : array, shape(n_features_out, n_features)
+            The solution computed by the optimizer in this iteration.
+        """
+        if self.callback is not None:
+            self.callback(transformation, self.n_iter_)
+
+        self.n_iter_ += 1
+
+    def _loss_grad_lbfgs(self, transformation, X, y, diffs,
+                         masks):
+        """Compute the loss and the loss gradient w.r.t. ``transformation``.
+
+        Parameters
+        ----------
+        transformation : array, shape (n_features_out, n_features)
+            The linear transformation on which to compute loss and evaluate
+            gradient
+        X : array, shape (n_samples, n_features)
+            The training samples.
+
+        y : array, shape (n_samples,)
+            The corresponding training labels.
+
+        diffs : array, shape (n_samples, n_samples, n_features)
+            Pairwise differences between training samples.
+
+        masks : array, shape (n_samples, n_classes)
+            One-hot encoding of y.
+
+        Returns
+        -------
+        loss : float
+            The loss computed for the given transformation.
+
+        gradient : array, shape (n_features_out * n_features,)
+            The new (flattened) gradient of the loss.
+        """
+
+        if self.n_iter_ == 0:
+            self.n_iter_ += 1
+            if self.verbose:
+                header_fields = ['Iteration', 'Objective Value', 'Time(s)']
+                header_fmt = '{:>10} {:>20} {:>10}'
+                header = header_fmt.format(*header_fields)
+                cls_name = self.__class__.__name__
+                print('[{}]'.format(cls_name))
+                print('[{}] {}\n[{}] {}'.format(cls_name, header,
+                                                cls_name, '-' * len(header)))
+
+        t_funcall = time.time()
+
+        transformation = transformation.reshape(-1, X.shape[1])
+        loss = 0
+        gradient = np.zeros(transformation.shape)
+        X_embedded = transformation.dot(X.T).T
+
+        # for every sample, compute its contribution to loss and gradient
+        for i in range(X.shape[0]):
+            diff_embedded = X_embedded[i] - X_embedded
+            sum_of_squares = np.einsum('ij,ij->i', diff_embedded,
```

Comment:
this variable contains distances (in embedded space), maybe it's good to have this in the variable name for clarity

Reply:
Done: I called them dist_embedded